Xplat/src/main/java/vazkii/botania/common/entity/ManaSparkEntity.java

```diff
@@ -163,29 +172,25 @@ public void tick() {
 
 			}
 			case DOMINANT -> {
-				List<ManaSpark> validSparks = SparkHelper.getSparksAround(level(), getX(), getY() + (getBbHeight() / 2), getZ(), getNetwork());
-				validSparks.removeIf(s -> {
-					SparkUpgradeType otherUpgrade = s.getUpgrade();
-					return s == this || otherUpgrade != SparkUpgradeType.NONE || !(s.getAttachedManaReceiver() instanceof ManaPool);
-				});
-				if (!validSparks.isEmpty()) {
-					validSparks.get(level().random.nextInt(validSparks.size())).registerTransfer(this);
+				if (receiverWasFull && !receiver.isFull()) {
+					updateTransfers();
+				}
+				if (!transfersTowardsSelfToRegister.isEmpty()) {
+					transfersTowardsSelfToRegister.remove(0).registerTransfer(this);
```

Comment:
If there's a lot of elements in the list, `remove(0)` is going to be expensive by forcing an arraylist recopy. Maybe you want a Deque instead?